heron/healthmgr/src/java/com/twitter/heron/healthmgr/sensors/MetricsCacheMetricsProvider.java

```diff
@@ -0,0 +1,162 @@
+    Map<String, ComponentMetrics> result = new HashMap<>();
+    for (String component : components) {
+      TopologyMaster.MetricResponse response =
+          getMetricsFromMetricsCache(metric, component, startTime, duration);
+      Map<String, InstanceMetrics> metrics = parse(response, component, metric);
+      ComponentMetrics componentMetric = new ComponentMetrics(component, metrics);
+      result.put(component, componentMetric);
+    }
+    return result;
+  }
+
+  @Override
+  public Map<String, ComponentMetrics> getComponentMetrics(String metric,
+                                                           Duration duration,
+                                                           String... components) {
+    Instant start = Instant.ofEpochMilli(clock.currentTime() - duration.toMillis());
+    return getComponentMetrics(metric, start, duration, components);
+  }
+
+  @VisibleForTesting
+  @SuppressWarnings("unchecked")
+  Map<String, InstanceMetrics> parse(
+      TopologyMaster.MetricResponse response, String component, String metric) {
+    Map<String, InstanceMetrics> metricsData = new HashMap<>();
+
+    if (response == null || !response.getStatus().getStatus().equals(StatusCode.OK)) {
+      return metricsData;
+    }
+
+    if (response.getMetricCount() == 0) {
+      LOG.info(String.format(
+          "Did not get any metrics from MetricsCache for %s:%s ", component, metric));
+      return metricsData;
+    }
+
+    // convert heron.protobuf.taskMetrics to dhalion.InstanceMetrics
+    for (TaskMetric tm :response.getMetricList()) {
+      String instanceId = tm.getInstanceId();
+      InstanceMetrics instanceMetrics = new InstanceMetrics(instanceId);
+
+      for (IndividualMetric im: tm.getMetricList()) {
+        String metricName = im.getName();
+        Map<Instant, Double> values = new HashMap<>();
+
+        for (IntervalValue iv: im.getIntervalValuesList()) {
+          MetricInterval mi = iv.getInterval();
+          String value = iv.getValue();
+          values.put(Instant.ofEpochSecond(mi.getStart()),  Double.parseDouble(value));
+        }
+
+        if (!values.isEmpty()) {
+          instanceMetrics.addMetric(metricName, values);
+        }
+      }
+
+      metricsData.put(instanceId, instanceMetrics);
+    }
+
+    return metricsData;
+  }
+
+  @VisibleForTesting
+  TopologyMaster.MetricResponse getMetricsFromMetricsCache(
+      String metric, String component, Instant start, Duration duration) {
+    TopologyMaster.MetricRequest request = TopologyMaster.MetricRequest.newBuilder()
+        .setComponentName(component)
+        .setExplicitInterval(
+            MetricInterval.newBuilder()
+            .setStart(start.getEpochSecond())
+            .setEnd(start.plus(duration).getEpochSecond())
+            .build())
+        .addMetric(metric)
+        .build();
+    LOG.fine("MetricsCache Query request: " + request.toString());
```

Comment:
In normal cases this string concatenation will not be needed and hence can be avoided. For e.g. using `LOG.log(Level.FINE, msg, param)`